sale_invoice_date_from_picking/models/sale_order.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2025 Quartile (https://www.quartile.co)
+# Licence AGPL-3.0 or later (https://www.gnu.org/licenses/agpl-3.0).
+
+from odoo import fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_picking_for_actual_date(self):
+        self.ensure_one()
+        return self.picking_ids.filtered(lambda p: p.state == "done").sorted(
+            key=lambda p: p.date_done, reverse=True
+        )[:1]
+
+    def _prepare_invoice(self):
+        invoice_vals = super()._prepare_invoice()
+        date_field = self.company_id.picking_date_field_for_invoice_date
+        if not date_field:
+            return invoice_vals
+        picking = self._get_picking_for_actual_date()
+        if not picking:
+            return invoice_vals
+        date_value = getattr(picking, date_field.name, False)
+        if date_field.ttype == "date" and date_value:
+            invoice_vals["invoice_date"] = date_value
+        else:
+            invoice_vals["invoice_date"] = fields.Datetime.context_timestamp(
+                self.with_context(tz=self.company_id.partner_id.tz),
+                date_value or picking.date_done,
+            )
```

Comment:
To be slightly safer.
```suggestion
        date_value = picking[date_field.name] or picking.date_done
        if date_field.ttype == "date":
            invoice_vals["invoice_date"] = value
        else:
            tz = self.company_id.partner_id.tz or self.env.user.tz
            invoice_vals["invoice_date"] = fields.Datetime.context_timestamp(self.with_context(tz=tz), value).date()
```

Reply:
As we talked, this will not respect the time of date_value if we choose date type field in the setting(eg.actual_date) and there is no value for this field in picking.

Reply:
@AungKoKoLin1997 We still need to let it fall back on user's timezone when company timezone is not available. Please take care of that.